File: WinForms/CS/GridMainDemo/CodeExamples/CellEditors.cs
using DevExpress.DXperience.Demos.CodeDemo;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Forms;
using System.ComponentModel;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraEditors.Controls;
using DevExpress.Data.Filtering;
using DevExpress.DXperience.Demos.CodeDemo.Data;
using SampleDataSourceFile = DevExpress.XtraGrid.Demos.CodeExamples.SampleDataCS;
using DevExpress.Utils;

namespace DevExpress.XtraGrid.Demos.CodeExamples {
    [CodeExampleClass("Cell editors", "CellEditors.cs")]
    public static class CellEditors {
        [CodeExampleSetUp]
        public static object[] SetUp(XtraUserControl sampleHost) {
            GridControl gridControl = new GridControl();
            GridView gridView = new GridView();
            gridControl.MainView = gridView;
            gridControl.Dock = DockStyle.Fill;
            gridControl.Parent = sampleHost;
            gridControl.DataSource = SampleData.GetData(10);
            gridView.PopulateColumns();

            return new object[] { gridControl, gridView };
        }
        [CodeExampleTearDown]
        public static void TearDown(XtraUserControl sampleHost) {
            (sampleHost.Controls[0] as GridControl).Dispose();
        }

        #region Cell editors

        [CodeExampleCase("Assign in-place editors dynamically"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("spinEdit", "calcEdit", "CustomRowCellEdit")]
        public static void AssignInPalceEditorsDynamically(GridControl gridControl, GridView gridView) {
            RepositoryItemSpinEdit spinEdit = new RepositoryItemSpinEdit();
            RepositoryItemCalcEdit calcEdit = new RepositoryItemCalcEdit();

            gridView.Columns["Length"].ShowButtonMode = ShowButtonModeEnum.ShowAlways;

            // Handle this event to assign editors to individual cells
            gridView.CustomRowCellEdit += (sender, e) => {
                GridView view = sender as GridView;
                if(e.Column.FieldName == "Length") {
                    bool boolVal = (bool)view.GetRowCellValue(e.RowHandle, "Mark");
                    if(boolVal)
                        e.RepositoryItem = spinEdit;
                    else
                        e.RepositoryItem = calcEdit;
                }
            };
        }

        [CodeExampleCase("Override the default in-place editor for certain cells"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("CustomRowCellEditForEditing", "gridLookUpEdit")]
        public static void OverrideTheDefaultInPlaceEditorForCertainCells(GridControl gridControl, GridView gridView) {
            RepositoryItemGridLookUpEdit gridLookUpEdit = new RepositoryItemGridLookUpEdit() {
                DisplayMember = "ID",
                ValueMember = "ID",
            };
            gridControl.RepositoryItems.Add(gridLookUpEdit);
            // The ID column uses a text editor by default.
            // The CustomRowCellEditForEditing event is handled to override this editor for rows. 
            gridView.CustomRowCellEditForEditing += (sender, e) => {
                if(e.Column.FieldName == "ID") {
                    // Assign data to gridLookUpEdit dynamically
                    gridLookUpEdit.DataSource = SampleData.GetData(10);
                    e.RepositoryItem = gridLookUpEdit;
                }
            };
            gridView.ShowEditor();
            (gridView.ActiveEditor as GridLookUpEdit).ShowPopup();
        }

        [CodeExampleCase("Assign in-place editors (repository items)"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("ColumnEdit")]
        public static void AssignInPlaceEditors(GridControl gridControl, GridView gridView) {
            // Assign a repository item to a column            
            RepositoryItemToggleSwitch edit = new RepositoryItemToggleSwitch();
            gridControl.RepositoryItems.Add(edit);
            gridView.Columns["Mark"].ColumnEdit = edit;
        }

        [CodeExampleCase("Show buttons in grid cells"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("ColumnEdit", "ShowButtonMode")]
        public static void ShowButtonsInGridCells(GridControl gridControl, GridView gridView) {
            // Show a button in a grid cell            
            RepositoryItemButtonEdit edit = new RepositoryItemButtonEdit();
            edit.ButtonClick += (sender, e) => {
                if(e.Button.Kind == ButtonPredefines.Delete) {
                    if(XtraMessageBox.Show("Do you wish to remove this row?", "Confirmation Dialog", MessageBoxButtons.YesNo) == DialogResult.Yes) {
                        gridView.DeleteRow(gridView.FocusedRowHandle);
                    }
                }
            };
            edit.Buttons[0].Kind = ButtonPredefines.Delete;
            edit.Buttons[0].Caption = "Custom Button";
            gridControl.RepositoryItems.Add(edit);
            GridColumn column = gridView.Columns["ID"];
            column.ColumnEdit = edit;
            column.ShowButtonMode = ShowButtonModeEnum.ShowAlways;
        }

        [CodeExampleCase("In-place mode immediate post changes"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("ColumnEdit")]
        public static void InplaceModeImmediatePostChanges(GridControl gridControl, GridView gridView) {
            // Immediate post changes
            RepositoryItemCheckEdit edit = new RepositoryItemCheckEdit() {
                InplaceModeImmediatePostChanges = DefaultBoolean.True
            };
            gridControl.RepositoryItems.Add(edit);
            gridView.Columns["Mark"].ColumnEdit = edit;

            GridColumn messageColumn = new GridColumn() {
                FieldName = "Message",
                Visible = true,
                UnboundDataType = typeof(string),
                UnboundExpression = "Iif([Mark], '', 'Not available in your region')",
            };
            gridView.Columns.Add(messageColumn);
        }

        #endregion
    }
}

File: WinForms/VB/GridMainDemo/CodeExamples/CellEditors.vb
Imports DevExpress.DXperience.Demos.CodeDemo
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Forms
Imports System.ComponentModel
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.Data.Filtering
Imports DevExpress.DXperience.Demos.CodeDemo.Data
Imports SampleDataSourceFile = DevExpress.XtraGrid.Demos.CodeExamples.SampleDataCS
Imports DevExpress.Utils

Namespace DevExpress.XtraGrid.Demos.CodeExamples

    <DevExpress.DXperience.Demos.CodeDemo.CodeExampleClass("Cell editors", "CellEditors.cs")>
    Public Module CellEditors

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleSetUp>
        Public Function SetUp(ByVal sampleHost As DevExpress.XtraEditors.XtraUserControl) As Object()
            Dim gridControl As DevExpress.XtraGrid.GridControl = New DevExpress.XtraGrid.GridControl()
            Dim gridView As DevExpress.XtraGrid.Views.Grid.GridView = New DevExpress.XtraGrid.Views.Grid.GridView()
            gridControl.MainView = gridView
            gridControl.Dock = System.Windows.Forms.DockStyle.Fill
            gridControl.Parent = sampleHost
            gridControl.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SampleData.GetData(10)
            gridView.PopulateColumns()
            Return New Object() {gridControl, gridView}
        End Function

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleTearDown>
        Public Sub TearDown(ByVal sampleHost As DevExpress.XtraEditors.XtraUserControl)
            TryCast(sampleHost.Controls(CInt((0))), DevExpress.XtraGrid.GridControl).Dispose()
        End Sub

#Region "Cell editors"
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Assign in-place editors dynamically"), DevExpress.XtraGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("spinEdit", "calcEdit", "CustomRowCellEdit")>
        Public Sub AssignInPalceEditorsDynamically(ByVal gridControl As DevExpress.XtraGrid.GridControl, ByVal gridView As DevExpress.XtraGrid.Views.Grid.GridView)
            Dim spinEdit As DevExpress.XtraEditors.Repository.RepositoryItemSpinEdit = New DevExpress.XtraEditors.Repository.RepositoryItemSpinEdit()
            Dim calcEdit As DevExpress.XtraEditors.Repository.RepositoryItemCalcEdit = New DevExpress.XtraEditors.Repository.RepositoryItemCalcEdit()
            gridView.Columns(CStr(("Length"))).ShowButtonMode = DevExpress.XtraGrid.Views.Base.ShowButtonModeEnum.ShowAlways
            ' Handle this event to assign editors to individual cells
            AddHandler gridView.CustomRowCellEdit, Sub(sender, e)
                Dim view As DevExpress.XtraGrid.Views.Grid.GridView = TryCast(sender, DevExpress.XtraGrid.Views.Grid.GridView)
                If Equals(e.Column.FieldName, "Length") Then
                    Dim boolVal As Boolean = CBool(view.GetRowCellValue(e.RowHandle, "Mark"))
                    If boolVal Then
                        e.RepositoryItem = spinEdit
                    Else
                        e.RepositoryItem = calcEdit
                    End If
                End If
            End Sub
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Override the default in-place editor for certain cells"), DevExpress.XtraGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("CustomRowCellEditForEditing", "gridLookUpEdit")>
        Public Sub OverrideTheDefaultInPlaceEditorForCertainCells(ByVal gridControl As DevExpress.XtraGrid.GridControl, ByVal gridView As DevExpress.XtraGrid.Views.Grid.GridView)
            Dim gridLookUpEdit As DevExpress.XtraEditors.Repository.RepositoryItemGridLookUpEdit = New DevExpress.XtraEditors.Repository.RepositoryItemGridLookUpEdit() With {.DisplayMember = "ID", .ValueMember = "ID"}
            gridControl.RepositoryItems.Add(gridLookUpEdit)
            ' The ID column uses a text editor by default.
            ' The CustomRowCellEditForEditing event is handled to override this editor for rows. 
            AddHandler gridView.CustomRowCellEditForEditing, Sub(sender, e)
                If Equals(e.Column.FieldName, "ID") Then
                    ' Assign data to gridLookUpEdit dynamically
                    gridLookUpEdit.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SampleData.GetData(10)
                    e.RepositoryItem = gridLookUpEdit
                End If
            End Sub
            gridView.ShowEditor()
            TryCast(gridView.ActiveEditor, DevExpress.XtraEditors.GridLookUpEdit).ShowPopup()
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Assign in-place editors (repository items)"), DevExpress.XtraGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("ColumnEdit")>
        Public Sub AssignInPlaceEditors(ByVal gridControl As DevExpress.XtraGrid.GridControl, ByVal gridView As DevExpress.XtraGrid.Views.Grid.GridView)
            ' Assign a repository item to a column            
            Dim edit As DevExpress.XtraEditors.Repository.RepositoryItemToggleSwitch = New DevExpress.XtraEditors.Repository.RepositoryItemToggleSwitch()
            gridControl.RepositoryItems.Add(edit)
            gridView.Columns(CStr(("Mark"))).ColumnEdit = edit
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Show buttons in grid cells"), DevExpress.XtraGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("ColumnEdit", "ShowButtonMode")>
        Public Sub ShowButtonsInGridCells(ByVal gridControl As DevExpress.XtraGrid.GridControl, ByVal gridView As DevExpress.XtraGrid.Views.Grid.GridView)
            ' Show a button in a grid cell            
            Dim edit As DevExpress.XtraEditors.Repository.RepositoryItemButtonEdit = New DevExpress.XtraEditors.Repository.RepositoryItemButtonEdit()
            AddHandler edit.ButtonClick, Sub(sender, e)
                If e.Button.Kind = DevExpress.XtraEditors.Controls.ButtonPredefines.Delete Then
                    If DevExpress.XtraEditors.XtraMessageBox.Show("Do you wish to remove this row?", "Confirmation Dialog", System.Windows.Forms.MessageBoxButtons.YesNo) = System.Windows.Forms.DialogResult.Yes Then
                        gridView.DeleteRow(gridView.FocusedRowHandle)
                    End If
                End If
            End Sub
            edit.Buttons(CInt((0))).Kind = DevExpress.XtraEditors.Controls.ButtonPredefines.Delete
            edit.Buttons(CInt((0))).Caption = "Custom Button"
            gridControl.RepositoryItems.Add(edit)
            Dim column As DevExpress.XtraGrid.Columns.GridColumn = gridView.Columns("ID")
            column.ColumnEdit = edit
            column.ShowButtonMode = DevExpress.XtraGrid.Views.Base.ShowButtonModeEnum.ShowAlways
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("In-place mode immediate post changes"), DevExpress.XtraGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("ColumnEdit")>
        Public Sub InplaceModeImmediatePostChanges(ByVal gridControl As DevExpress.XtraGrid.GridControl, ByVal gridView As DevExpress.XtraGrid.Views.Grid.GridView)
            ' Immediate post changes
            Dim edit As DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit = New DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit() With {.InplaceModeImmediatePostChanges = DevExpress.Utils.DefaultBoolean.[True]}
            gridControl.RepositoryItems.Add(edit)
            gridView.Columns(CStr(("Mark"))).ColumnEdit = edit
            Dim messageColumn As DevExpress.XtraGrid.Columns.GridColumn = New DevExpress.XtraGrid.Columns.GridColumn() With {.FieldName = "Message", .Visible = True, .UnboundDataType = GetType(String), .UnboundExpression = "Iif([Mark], '', 'Not available in your region')"}
            gridView.Columns.Add(messageColumn)
        End Sub
#End Region
    End Module
End Namespace
